### PR TITLE
[ES-86] Add TLS configuration information.

### PR DESCRIPTION
**Pull request by bgravenorst**
Signed-off-by: Byron Gravenorst <byron.gravenorst@consensys.net>

## Pull Request Description
Add information about how to configure TLS.

## Fixed Issue(s)
ES-48

### CODE REVIEW

File: docs/HowTo/Configure-TLS.md [hunk 1]
```diff
@@ -0,0 +1,118 @@
...
+* Specifies the EthSigner keystore using the [`--tls-keystore-file`](../Reference/CLI/CLI-Syntax.md#tls-keystore-file) option.
+* Specifies the file that contains the password to decrypt the
+keystore using the [`--tls-keystore-password-file`](../Reference/CLI/CLI-Syntax.md#tls-keystore-password-file) option.
+* [Specify the clients](#create-the-known-clients-file) that are trusted to connect to EthSigner using the
+[`tls-known-clients-file`](../Reference/CLI/CLI-Syntax.md#tls-known-clients-file) option.
+* Allow access to clients with trusted CA certificates using the [`--tls-allow-ca-clients`](../Reference/CLI/CLI-Syntax.md#tls-allow-ca-clients) option.
+
+!!! note
+    Use the [`--tls-allow-any-client`](../Reference/CLI/CLI-Syntax.md#tls-allow-any-client) option to allow
+    access to any client that presents a certificate.
```
Comment (usmansaleem) [inline docs/HowTo/Configure-TLS.md]: @rain-on can confirm, but I think when specifying `--tls-allow-any-client`, the client doesn't need to present a certificate at all (i.e. we disable client auth). If it presents, I believe we ignore it if this option is set.

File: docs/HowTo/Configure-TLS.md [hunk 2]
```diff
@@ -0,0 +1,118 @@
...
+* Enables TLS using the [`--downstream-http-tls-enabled`](../Reference/CLI/CLI-Syntax.md#downstream-http-tls-enabled)
+option.
+* Specifies the keystore to present during authentication using the 
+[`--downstream-http-tls-keystore-file`](../Reference/CLI/CLI-Syntax.md#downstream-http-tls-keystore-file) option.
+* Specifies the file that contains the password to decrypt the keystore using
+the [`--downstream-http-tls-keystore-password-file`](../Reference/CLI/CLI-Syntax.md#downstream-http-tls-keystore-password-file) option.
+* [Specifies the servers](#create-the-known-servers-file) to connect to using the [`--downstream-http-tls-known-servers-file`](../Reference/CLI/CLI-Syntax.md#downstream-http-tls-known-servers-file) option.
+
+!!! note
+    The [`downstream-http-tls-ca-auth-enabled`](../Reference/CLI/CLI-Syntax.md#downstream-http-tls-ca-auth-enabled) option is enabled
```
Comment (usmansaleem) [inline docs/HowTo/Configure-TLS.md]: does `downstream-http-tls-ca-auth-enabled` needs to be preceded with `--`?

File: docs/Reference/CLI/CLI-Syntax.md [hunk 3]
```diff
@@ -89,6 +89,72 @@ Timeout period (in milliseconds) for downstream requests. Default is 5000.
...
+--downstream-http-tls-ca-auth-enabled[=<true|false>]
+```
+
+```bash tab="Example"
+--downstream-http-tls-enabled=false
+```
+
+### downstream-http-tls-keystore-file
+
+Keystore file that is presented to the server during authentication.
```
Comment (usmansaleem) [inline docs/Reference/CLI/CLI-Syntax.md]: It should be rephrased something similar to 

> Keystore file (in PKCS #12 format) that contains private key and the certificate presented to the server during authentication.

File: docs/Reference/CLI/CLI-Syntax.md [hunk 4]
```diff
@@ -134,6 +200,61 @@ Displays the help and exits.
 -h, --help
 ```
 
+### tls-allow-any-client
+
+Allows any client that presents a certificate to connect.
```
Comment (usmansaleem) [inline docs/Reference/CLI/CLI-Syntax.md]: @rain-on please re-confirm. I think with `tls-allow-any-client`, the client may not have to present any certificate at all.

File: docs/Reference/CLI/CLI-Syntax.md [hunk 5]
```diff
@@ -134,6 +200,61 @@ Displays the help and exits.
...
+!!! important 
+    Cannot be used with `--tls-allow-ca-clients` and `--tls-known-clients-file`
+    
+```bash tab="Syntax"
+--tls-allow-any-client
+```
+
+### tls-allow-ca-clients
+
+Allows clients with trusted CA certificates to connect.
```
Comment (usmansaleem) [inline docs/Reference/CLI/CLI-Syntax.md]: > Allows clients signed with trust CA certificates to connect.

File: docs/HowTo/Configure-TLS.md (hunk 1 above)
Comment (bgravenorst) [reply docs/HowTo/Configure-TLS.md]: Confirmed with Trent, the client does not need to present a certificate.

File: docs/HowTo/Configure-TLS.md (hunk 2 above)
Comment (bgravenorst) [reply docs/HowTo/Configure-TLS.md]: Done

File: docs/Reference/CLI/CLI-Syntax.md (hunk 3 above)
Comment (bgravenorst) [reply docs/Reference/CLI/CLI-Syntax.md]: Done.

File: docs/Reference/CLI/CLI-Syntax.md (hunk 4 above)
Comment (bgravenorst) [reply docs/Reference/CLI/CLI-Syntax.md]: Done.

File: docs/Reference/CLI/CLI-Syntax.md (hunk 5 above)
Comment (bgravenorst) [reply docs/Reference/CLI/CLI-Syntax.md]: Done

File: docs/HowTo/Configure-TLS.md [hunk 6]
```diff
@@ -0,0 +1,118 @@
...
+* Enables TLS using the [`--downstream-http-tls-enabled`](../Reference/CLI/CLI-Syntax.md#downstream-http-tls-enabled)
+option.
+* Specifies the keystore to present during authentication using the 
+[`--downstream-http-tls-keystore-file`](../Reference/CLI/CLI-Syntax.md#downstream-http-tls-keystore-file) option.
+* Specifies the file that contains the password to decrypt the keystore using
+the [`--downstream-http-tls-keystore-password-file`](../Reference/CLI/CLI-Syntax.md#downstream-http-tls-keystore-password-file) option.
+* [Specifies the servers](#create-the-known-servers-file) to connect to using the [`--downstream-http-tls-known-servers-file`](../Reference/CLI/CLI-Syntax.md#downstream-http-tls-known-servers-file) option.
+
+!!! note
+    The [`--downstream-http-tls-ca-auth-enabled`](../Reference/CLI/CLI-Syntax.md#downstream-http-tls-ca-auth-enabled) option is enabled
```
Comment (MadelineMurray) [inline docs/HowTo/Configure-TLS.md]: is enabled -> is true?

File: docs/HowTo/Store-Keys/Use-Hashicorp.md [hunk 7]
```diff
@@ -54,15 +52,32 @@ option set to `8590` to avoid conflict with the default EthSigner listening port
...
 
 Start EthSigner.
 
 !!! example
 
     ```bash
-    ethsigner --chain-id=2018 --downstream-http-port=8590 hashicorp-signer --host=127.0.0.1 --port=8200 --auth-file=authFile
+    ethsigner --chain-id=2018 --downstream-http-port=8590 hashicorp-signer --host=127.0.0.1 --port=8200 --auth-file=authFile --tls-known-server-file=/Users/me/my_node/knownHashicorpServers
     ```
+To disable TLS set [`hashicorp-signer --tls-enabled`](../../Reference/CLI/CLI-Syntax.md#tls-enabled) to `false`.
```
Comment (MadelineMurray) [inline docs/HowTo/Store-Keys/Use-Hashicorp.md]: To disable TLS set -> To disable TLS, set

File: docs/HowTo/Store-Keys/Use-Hashicorp.md [hunk 8]
```diff
@@ -54,15 +52,32 @@ option set to `8590` to avoid conflict with the default EthSigner listening port
     protection. That is, the genesis file must include at least the Spurious Dragon milestone
     (defined as `eip158Block` in the genesis file) so the blockchain is using a chain ID.
 
+## Create the Known Servers File
```
Comment (MadelineMurray) [inline docs/HowTo/Store-Keys/Use-Hashicorp.md]: Given TLS is enabled by default, is it a requirement to create this file unless you explicitly disable TLS?  ie, what happens if you don't create it and haven't explicitly disabled TLS? 

If it's a requirement to create this file when TLS is on by default, it probably makes sense to move the sentence from the following section about how to disable TLS to this section.  And also make it clear creating this file is required unless you explicitly disable TLS.

File: docs/Reference/CLI/CLI-Syntax.md [hunk 9]
```diff
@@ -89,6 +89,70 @@ Timeout period (in milliseconds) for downstream requests. Default is 5000.
 --downstream-http-request-timeout=3000
 ```
 
+### downstream-http-tls-enabled
+
+Enable or disable [TLS for server connections](../../Concepts/TLS.md).
+
+```bash tab="Syntax"
+--downstream-http-tls-enabled
```
Comment (MadelineMurray) [inline docs/Reference/CLI/CLI-Syntax.md]: Needs the `[=<true|false>]` I assume?  ie, how to you disable this?

Comment (MadelineMurray) [reply docs/Reference/CLI/CLI-Syntax.md]: What is the default?

File: docs/Concepts/TLS.md [hunk 10]
```diff
@@ -0,0 +1,18 @@
...
+
+![EthSigner TLS](../images/Ethsigner_TLS.png)
+
+
+Private keys and certificates for client and server TLS connections must be
+stored in password-protected PKCS #12 keystores. The server 
+(in this example, Hyperledger Besu) must be configured to accept TLS
+connections.
+
+Use the command line options to [enable and configure](../HowTo/Configure-TLS.md) TLS.
```
Comment (MadelineMurray) [inline docs/Concepts/TLS.md]: This sentence only links to the content on enabling and configuring TLS on the HTTP JSON-RPC connection.  Is it worth also linking the the configuring Hashicorp section as well so it's clear there's two places that need configured?

File: docs/Concepts/TLS.md [hunk 11]
```diff
@@ -0,0 +1,18 @@
...
+# TLS Communication
+
+EthSigner supports TLS to secure inbound and outbound HTTP JSON-RPC
+requests, and [communication with Hashicorp Vault](../HowTo/Store-Keys/Use-Hashicorp.md).
+
+![EthSigner TLS](../images/Ethsigner_TLS.png)
+
+
+Private keys and certificates for client and server TLS connections must be
+stored in password-protected PKCS #12 keystores. The server
```
[the remark this answers is not in the document]
Comment (bgravenorst) [reply docs/Concepts/TLS.md]: As discussed, if they're not sure, there's plenty of resources available.

File: docs/HowTo/Configure-TLS.md [hunk 12]
```diff
@@ -0,0 +1,118 @@
+---
+description: Secure communication using TLS
+---
+
+# Configure TLS
+
+Configure TLS communication from the command line. Clients and servers
+communicate using HTTP JSON-RPCs.
+
+**Prerequisites**:
```
[the remark this answers is not in the document]
Comment (bgravenorst) [reply docs/HowTo/Configure-TLS.md]: Done

File: docs/HowTo/Configure-TLS.md [hunk 13]
```diff
@@ -0,0 +1,118 @@
...
+
+* `<common_name>` is the Common Name used for the client's keystore
+* `<hex-string>` is the SHA-256 fingerprint of the client's keystore.
+
+!!! example
+    ```
+    curl_client DF:65:B8:02:08:5E:91:82:0F:91:F5:1C:96:56:92:C4:1A:F6:C6:27:FD:6C:FC:31:F2:BB:90:17:22:59:5B:50
+    ```
+    
+You can use openssl or keytool to display the fingerprint. For example:
```
[the remark this answers is not in the document]
Comment (bgravenorst) [reply docs/HowTo/Configure-TLS.md]: Done.

File: docs/HowTo/Configure-TLS.md [hunk 14]
```diff
@@ -0,0 +1,118 @@
...
+```bash
+ethsigner --tls-keystore-file=/Users/me/my_node/keystore.pfx --tls-keystore-password-file=/Users/me/my_node/keystorePassword --tls-known-clients-file=/Users/me/my_node/knownClients  --tls-allow-ca-clients
+```
+
+The command line:
+
+* Specifies the EthSigner keystore using the [`--tls-keystore-file`](../Reference/CLI/CLI-Syntax.md#tls-keystore-file) option.
+* Specifies the file that contains the password to decrypt the
+keystore using the [`--tls-keystore-password-file`](../Reference/CLI/CLI-Syntax.md#tls-keystore-password-file) option.
+* [Specify the clients](#create-the-known-clients-file) that are trusted to connect to EthSigner using the
```
[the remark this answers is not in the document]
Comment (bgravenorst) [reply docs/HowTo/Configure-TLS.md]: Done

File: docs/HowTo/Configure-TLS.md (hunk 6 above)
Comment (bgravenorst) [reply docs/HowTo/Configure-TLS.md]: Done

File: docs/HowTo/Store-Keys/Use-Hashicorp.md (hunk 7 above)
Comment (bgravenorst) [reply docs/HowTo/Store-Keys/Use-Hashicorp.md]: Done

File: docs/HowTo/Store-Keys/Use-Hashicorp.md (hunk 8 above)
Comment (bgravenorst) [reply docs/HowTo/Store-Keys/Use-Hashicorp.md]: Done.

File: docs/Reference/CLI/CLI-Syntax.md (hunk 9 above)
Comment (bgravenorst) [reply docs/Reference/CLI/CLI-Syntax.md]: Done.

File: docs/Concepts/TLS.md (hunk 10 above)
Comment (bgravenorst) [reply docs/Concepts/TLS.md]: Done.

**Pull request merged**